crux_core/src/capability.rs

```diff
@@ -266,37 +266,92 @@ where
 /// }
 /// ```
 ///
-pub struct CapabilityContext<Ef, Event> {
-    inner: std::sync::Arc<ContextInner<Ef, Event>>,
+pub struct CapabilityContext<Op, Event>
+where
+    Op: Operation,
+{
+    inner: std::sync::Arc<ContextInner<Op, Event>>,
+}
+
+struct ContextInner<Op, Event>
+where
+    Op: Operation,
+{
+    shell_channel: Sender<Step<Op>>,
+    app_channel: Sender<Event>,
+    spawner: crate::executor::Spawner,
 }
 
-struct ContextInner<Ef, Event> {
-    steps: Sender<Step<Ef>>,
-    events: Sender<Event>,
+pub struct ProtoContext<E, Event> {
+    shell_channel: Sender<E>,
+    app_channel: Sender<Event>,
     spawner: crate::executor::Spawner,
 }
 
-impl<Ef, Ev> Clone for CapabilityContext<Ef, Ev> {
+impl<Op, Ev> Clone for CapabilityContext<Op, Ev>
+where
+    Op: Operation,
+{
     fn clone(&self) -> Self {
         Self {
             inner: Arc::clone(&self.inner),
         }
     }
 }
 
-impl<T, Ev> CapabilityContext<T, Ev>
+impl<E, Ev> ProtoContext<E, Ev>
 where
-    T: Send + 'static,
     Ev: 'static,
+    E: 'static,
 {
     pub(crate) fn new(
-        steps: Sender<Step<T>>,
+        steps: Sender<E>,
         events: Sender<Ev>,
         spawner: crate::executor::Spawner,
     ) -> Self {
+        Self {
+            shell_channel: steps,
+            app_channel: events,
+            spawner,
+        }
+    }
+
+    /// Specialise the CapabilityContext to a specific capability, wrapping its operations into
+    /// an Effect `Ef`. The `func` argument will typically be an Effect variant constructor, but
+    /// can be any function taking the capability's operation type and returning
+    /// the effect type.
+    ///
+    /// This will likely only be called from the implementation of [`WithContext`]
+    /// for the app's `Capabilities` type.
+    pub fn specialise<Op, F>(&self, func: F) -> CapabilityContext<Op, Ev>
+    where
+        F: Fn(Step<Op>) -> E + Sync + Send + Copy + 'static,
+        Op: Operation,
+    {
         let inner = Arc::new(ContextInner {
-            steps,
-            events,
+            shell_channel: self.shell_channel.map_input(func),
+            app_channel: self.app_channel.clone(),
+            spawner: self.spawner.clone(),
+        });
+
+        CapabilityContext { inner }
```

Comment:
Could get rid of the `#[cfg(test)]` on line 346 and just call `CapabilityContext::new(...)` here.